### PR TITLE
Dedicated and Shared Worker Registry

### PR DESCRIPTION
**Pull request by m-cheung**
Please do not merge - Work in progress.
Input / suggestions requested.

### CODE REVIEW

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  requires: [
+    'foam.core.StubFactorySingleton',
+    'foam.box.BoxRegistry',
+    'foam.box.node.ForkBox'
+  ],
+
+  imports: [ 'registry' ], // __context__.registry is the context it is declared in
+
+  exports: [ 'as registry' ], // __subContext__.registry === this
+
```
Comment (m-cheung) [inline src/foam/box/DedicatedWorkerRegistry.js]: I am not sure if this is the appropriate action. Previous, it was discussed that we would do something along the lines of
```
var ctx = foam.box.Context();
ctx.registry = foam.box.DedicatedWorkerRegistry.create({
  delegate: ctx.registry
}, ctx);
```
But this causes a cyclic reference to itself. If we create them separately
```
var ctx = foam.box.Context();
var registry = foam.box.DedicatedWorkerRegistry.create({
  delegate: ctx.registry
}, ctx);
```
then use `registry.register`, registrations will be ok, but issues will come up later on. An example:
```
var stub = foam.core.StubFactorySingleton.create().get(foam.dao.DAO)
   .create({ delegate: register }, ctx);
stub.put(...)
```
As this will register will a different registry, and then the stub call will not be able to find the box in the registry, (in the current implementation)

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        stubFactory: function() {
+          return self.stubFactory_.create({
+            delegate: self.workerFactory(this)
+          }, this);
+        },
+
+        registerWorker: function(name, service, box) {
+          return self.dedicatedWorkers_[key].register(name, service, box);
+        }
+      });
```
Comment (m-cheung) [inline src/foam/box/DedicatedWorkerRegistry.js]: Perhaps, line 64-74 is where my misinterpretation begins? I don't think you intended me to handle the context's in this sense.

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.box',
+  name: 'DedicatedWorkerRegistry',
+  extends: 'foam.box.ProxyBox',
+  implements: [ 'foam.box.BoxRegistryBox' ],
```
Comment (m-cheung) [inline src/foam/box/DedicatedWorkerRegistry.js]: I naturally thought that this is more of an extension of `foam.box.BoxRegistryBox` and it implements `foam.box.ProxyBox` but doing so is not possible with importing registry as well.

(Additionally, most places tend to extend `foam.box.ProxyBox`, so I believe this should be the right way of doing this)

File: src/foam/box/SharedWorkerRegistry.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
+/**
```
Comment (m-cheung) [inline src/foam/box/SharedWorkerRegistry.js:1]: This file can be ignored for now. It will be updated once `DedicatedWorkerRegistry` is fixed.

File: test/src/box/DedicatedWorkerRegistry.js [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+    MockRegistry = foam.lookup('foam.box.DedicatedWorkerRegistry.Test.MockRegistry');
+    Registry = foam.lookup('foam.box.DedicatedWorkerRegistry');
+    Context = foam.lookup('foam.box.Context');
+    LogBox = foam.lookup('foam.box.LogBox');
+  });
+
+  it('should forward all registrations to delegate by default', function() {
+    var ctx = Context.create();
+    var mockRegistry = MockRegistry.create(null, ctx);
```
Comment (m-cheung) [inline test/src/box/DedicatedWorkerRegistry.js]: Assume that `mockRegistry` is acting as the "default" registry throughout these tests

File: test/src/box/DedicatedWorkerRegistry.js [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+    ctx.registry = Registry.create({
+      delegate: mockRegistry
+    }, ctx);
```
Comment (m-cheung) [inline test/src/box/DedicatedWorkerRegistry.js]: Is this how we intend on using `DedicatedWorkerRegistry`? Or do we intend to use the registry separately from the context?

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+      this.register = oldReg;
+      return ret;
+    },
+    function unregister(name) {
+      // Attempt to unregister in ourselves first.
+      if ( foam.box.Box.isInstance(name) ) {
+        var key = this.getDedicatedWorkerKey(name);
+
+        for (var key in this.dedicatedWorkers_ ) {
+          if ( this.dedicatedWorkers_[key] === name ) {
```
Comment (m-cheung) [inline src/foam/box/DedicatedWorkerRegistry.js]: Do we need to worry about removing `StubFactory<class>` from `dedicatedWorkers_`?
If not, we can simply this code

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.box',
+  name: 'DedicatedWorkerRegistry',
+  extends: 'foam.box.BoxRegistryBox', //'foam.box.ProxyBox',
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: nit: Here and elsewhere: Drop commented-out code.

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    'foam.box.node.ForkBox'
+  ],
+
+//  imports: [ 'registry' ], // __context__.registry is the context it is declared in
+
+  exports: [ 'as registry' ], // __subContext__.registry === this
+
+  properties: [
+    {
+      name: "getDedicatedWorkerKey",
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: `class: 'Function'`. You may need to use a `factory` instead of `value` if `foam.core.Function` has a factory.

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+      value: function(box) { return null; }
+    },
+    {
+      name: 'workerFactory',
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: `class: 'Function',`

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    'foam.core.StubFactorySingleton',
+    'foam.box.BoxRegistry',
+    'foam.box.node.ForkBox'
+  ],
+
+//  imports: [ 'registry' ], // __context__.registry is the context it is declared in
+
+  exports: [ 'as registry' ], // __subContext__.registry === this
+
+  properties: [
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: Per-property documentation strings, please :smile:

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+foam.CLASS({
+  package: 'foam.box',
+  name: 'DedicatedWorkerRegistry',
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js:20]: Documentation with example, please 😸

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+      factory: function() {
+        return this.StubFactorySingleton.create().get(this.BoxRegistry);
+      }
+    }
+  ],
+
+  methods: [
+    function register(name, service, box) {
+      var key = this.getDedicatedWorkerKey(box);
+      if ( ! key ) return this.delegate.register(name, service, box);
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: if `delegate` isn't marked as `required: true` in `ProxyBox`, then we should add that in this class:

```js
properties: [
  {
    name: 'delegate',
    documentation: `...`,
    required: true
  }
]
```

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+      // Redirecting all registers of following statements with our registry.
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: Wait, what?

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+  exports: [ 'as registry' ], // __subContext__.registry === this
+
+  properties: [
+    {
+      name: "getDedicatedWorkerKey",
+      value: function(box) { return null; }
+    },
+    {
+      name: 'workerFactory',
+      value: function(ctx) {
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: Let's make this receive `(args, ctx)`, like the `create` interface.

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+      // Redirecting all registers of following statements with our registry.
+      // If this is not done, this register method will be called recursively,
+      // and registrations will be forwarded to delegate.
+      var oldReg = this.register;
+      this.register = this.SUPER;
+
+      if ( ! this.dedicatedWorkers_[key] )
+        this.dedicatedWorkers_[key] = this.stubFactory_.create({
+          delegate: this.workerFactory(this)
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: Shouldn't `(this)` be `(this.__context__)` to avoid the aforementioned recursion? Such a strategy instantiates `ForkBox` (or whatever `workerFactory` is injected to do) in a context that has `this`'s `registry`, not `this`'s subcomponents' registry.

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+  methods: [
+    function register(name, service, box) {
+      var key = this.getDedicatedWorkerKey(box);
+      if ( ! key ) return this.delegate.register(name, service, box);
+
+      // Redirecting all registers of following statements with our registry.
+      // If this is not done, this register method will be called recursively,
+      // and registrations will be forwarded to delegate.
+      var oldReg = this.register;
+      this.register = this.SUPER;
```
Comment (m-cheung) [inline src/foam/box/DedicatedWorkerRegistry.js]: Is there a safer way of doing this? The following function needs to run in the same context, except with a different register method. I tried using subContext's but I could not find a way to change just the method

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+      var ret = this.dedicatedWorkers_[key].register(name, service, box);
+
+      // Register is set back to this method for future calls.
+      this.register = oldReg;
+      return ret;
+    },
+    function unregister(name) {
+      // Attempt to unregister in ourselves first.
+      if ( foam.box.Box.isInstance(name) ) {
+        var key = this.getDedicatedWorkerKey(name);
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: I don't think this will work reliably. `name` here is the box returned above (`ret`), not the box passed in to `register()` above.

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+        for (var key in this.dedicatedWorkers_ ) {
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: ditto.

File: src/foam/box/DedicatedWorkerRegistry.js (hunk 7 above)
Comment (mdittmer) [reply src/foam/box/DedicatedWorkerRegistry.js]: To do this correctly we would need to keep a local collection of services (still) registered on this dedicated worker. When the size of the collection reaches zero, then we could delete this. For now, I'd say it can be a `TODO` to clean up dedicated workers.

**Review by mdittmer — CHANGES_REQUESTED** [review]
Reviewed `foam.box.DedicatedWorkerRegistry`. In addition to inline comments, `unregister()` will be unsound without additional state tracking.

`register()` needs mutate some kind of  bookkeeping that associates workers with names and/or exported boxes (`ret` in `register()`). Getting the name when name was unspecified is a an edge case here. Another edge case is when the return value from `dedicatedWorker_[key].register(...)` has a `PromisedBox` in its delegate chain (so its `Promise` could reject, and that aforementioned bookkeeping needs to be cleaned up). Then, `unregister()` can use the bookkeeping to determine whether or not it should handle the call or delegate. It should never do both.

We can chat about this offline if you like.

File: src/foam/box/DedicatedWorkerRegistry.js (hunk 16 above)
Comment (m-cheung) [reply src/foam/box/DedicatedWorkerRegistry.js]: I may still have something wired incorrectly then. I checked both `this.__context__.registry` and `this.__subContext.registry` and they are both referring to `this` (i.e. `this.__context__.registry === this`)

File: src/foam/box/DedicatedWorkerRegistry.js (hunk 18 above)
Comment (m-cheung) [reply src/foam/box/DedicatedWorkerRegistry.js]: I have actually simplified this to just
```
this.SUPER(name); 
this.delegate.unregister(name);
```

This does assume that name of the box is unique (I believe this is true, but I may have to add a check on register) and should only be registered in one of the registries (this should be true).

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 20]
```diff
@@ -18,36 +18,81 @@
...
 
-//  imports: [ 'registry' ], // __context__.registry is the context it is declared in
-
-  exports: [ 'as registry' ], // __subContext__.registry === this
+  exports: [ 'as registry' ],
 
   properties: [
     {
+      class: 'Function',
       name: "getDedicatedWorkerKey",
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: nit: single quotes

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 21]
```diff
@@ -18,36 +18,81 @@
...
-      value: function(box) { return null; }
+      documentation: `Gets the service name from the provided box.
+        If the value is not null, a dedicated worker will be used for the box.
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: nit: +4 indent after first line.

File: src/foam/box/DedicatedWorkerRegistry.js [hunk 22]
```diff
@@ -57,7 +102,9 @@ foam.CLASS({
   methods: [
     function register(name, service, box) {
       var key = this.getDedicatedWorkerKey(box);
-      if ( ! key ) return this.delegate.register(name, service, box);
+      if ( ! key ) {
+        return this.delegate.register(name, service, box);
+      }
```
Comment (mdittmer) [inline src/foam/box/DedicatedWorkerRegistry.js]: Lines 109-114, 121-123 should be gone now, right?

File: test/src/box/DedicatedWorkerRegistry.js (hunk 6 above)
Comment (mdittmer) [reply test/src/box/DedicatedWorkerRegistry.js]: It should be usable anywhere you would use a `BoxRegistry`. In practice, it's likely that a `PipelineBuilder` will have `imports: [ 'registry' ]` which will probably contain `DedicatedWorkerRegistry -> SharedWorkerRegistry`, possibly with `-> BoxRegistry` on the end.

File: test/src/box/DedicatedWorkerRegistry.js [hunk 23]
```diff
@@ -70,10 +70,14 @@ describe('dedicated worker registry', function() {
       delegate: mockRegistry
     }, ctx);
 
+    /*
```
Comment (mdittmer) [inline test/src/box/DedicatedWorkerRegistry.js]: nit: Clean this up once test code is as you want it.

File: src/foam/box/BoxRegistry.js [hunk 24]
```diff
@@ -70,6 +70,13 @@ foam.CLASS({
       ]
     },
     {
+      name: 'getRegisteredNames',
+      returns: 'Array',
+      code: function() {
+        return Object.keys(this.registry);
+      }
+    },
+    {
```
Comment (m-cheung) [inline src/foam/box/BoxRegistry.js]: Having `DedicatedWorkerRegistry` keep track of registrations is not sufficient. This is because we pass `localRegistry` to functions, which make registrations that do not go through `DedicatedWorkerRegistry` (but technically belonging to `DedicatedWorkerRegistry`).

Thus, name conflicts would still be possible if we only relied on a map property that is updated upon register and unregister.

File: src/foam/box/BoxRegistryBox.js [hunk 25]
```diff
@@ -45,7 +45,7 @@ foam.CLASS({
         if ( this.SubBoxMessage.isInstance(msg.object) ) {
           var name = msg.object.name;
 
-          if ( this.registry[name].localBox ) {
+          if ( this.registry[name] && this.registry[name].localBox ) {
```
Comment (m-cheung) [inline src/foam/box/BoxRegistryBox.js:48]: Previous check was insufficient. If `this.registry[name]` is not defined, it will not move on to attempting to send an error back through `msg.attributes.errorBox`

File: src/foam/box/DedicatedWorkerRegistry.js (hunk 12 above)
Comment (m-cheung) [reply src/foam/box/DedicatedWorkerRegistry.js:20]: Addressed below.

File: test/src/box/DedicatedWorkerRegistry.js [hunk 26]
```diff
@@ -0,0 +1,256 @@
...
+    var dao = foam.dao.ArrayDAO.create({ of: 'Object' });
+    var box = foam.box.SkeletonBox.create({ data: dao });
+    box.serviceName = "ArrayDAO"; // Enabling dedicated service worker.
+
+    // Perform remote registration of service.
+    var register = ctx.registry.register(null, null, box);
+
+    // Creating a stub on local for accessing method.
+    var stub = foam.core.StubFactorySingleton.create().get(foam.dao.DAO)
+        .create({ delegate: register }, ctx.registry.localRegistry);
```
Comment (m-cheung) [inline test/src/box/DedicatedWorkerRegistry.js:135]: Should the user need to know which registry the box above was registered in? If `ctx` is provided instead (i.e. allowing `DedicatedWorkerRegistry` to handle the registration), it will perform registrations for the stub in the delegate's registry. This results in issues when we call `stub.put(obj)` below (it will not be able to find the ReplyBox).

A stack trace captured at the lookup failure:
```
code | @ | BoxRegistryBox.js:48
send | @ | DedicatedWorkerRegistry.js:146
send | @ | NamedBox.js:53
code | @ | SubBox.js:42
(anonymous function) | @ | Promised.js:56
_tickCallback | @ | internal/process/next_tick.js:109
_Promise.resolve (async)_
(anonymous function) | @ | ForkBox.js:81
send | @ | ReplyBox.js:42
code | @ | BoxRegistryBox.js:51
(anonymous function) | @ | net.js:471
```

To be clear: providing `ctx.registry.localRegistry` will work fine. I'm just not sure if the user should be required to specify the correct registry.

**Comment by m-cheung** [comment]
@mdittmer I noticed while writing my tests that processes spawned by ForkScript are not killed if the exit status is 0. When running Jasmine tests, the child processes will continue to run after, even when all of the tests have completed and Jasmine / Node has exited.

The child processes are killed if we send a kill signal (e.g. `Ctrl`+`C` or `SIGTERM`) is sent, but that is about it. Here is a snippet of code you can use to reproduce this behaviour.

```
require('./src/foam.js');                                                                                              
                                                                                                                                                                                                         
var ctx = foam.box.Context.create();                                                                                   
var remoteRegistry = stubFactory.get(foam.box.BoxRegistry).create({                                                    
  delegate: foam.box.node.ForkBox.create(null, ctx)                                                                    
}, ctx);                                                                                                               
                                                                                                                       
var serviceBox = remoteRegistry.register(null, null, foam.box.LogBox.create());                                                                     
serviceBox.send(foam.box.Message.create({ object: 'Hello world' }));

setTimeout(function() {
  process.exit();
}, 2000);
```

TLDR: I think we need a mechanism to clean up child processes in ForkBox

**Comment by m-cheung** [comment]
Issues were addressed in #547
